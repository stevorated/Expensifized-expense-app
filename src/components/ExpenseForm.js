import React from 'react'

import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'


export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props)

        this.state = {
            desc: props.expense ? props.expense.desc : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount/100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            // calanderFocused: false,
            error: ''
    
        }
    }

    onDescriptionChange = (e) => {
        const desc = e.target.value
        this.setState(()=> ({
            desc
        }))
    }

    onAmountChange = (e) => {
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({
                amount
            }))
        }
        
    }
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(()=>({createdAt}))
        }
    }

    // onFocusChange = (calanderFocused) => {
    //     this.setState(()=> ({calanderFocused}))
    // }

    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(()=> ({
            note
        }))
    }

    onSubmit = (e) => {
        e.preventDefault()

        if(!this.state.desc || !this.state.amount) {
            const error = 'Please provide description and amount to save.'
            this.setState(()=>({
                error
            })  )
        } else {
            this.setState(()=> ({
                error: ''
            }))
            this.props.onSubmit({
                desc: this.state.desc,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    {this.state.error && <p>{this.state.error}</p>}
                    <input 
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.desc}
                    onChange={this.onDescriptionChange}
                    />
                    <input 
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null
                    onDateChange={createdAt => this.setState({ createdAt })} // PropTypes.func.isRequired
                    focused={this.state.focused} // PropTypes.bool
                    onFocusChange={({ focused }) => this.setState({ focused })} // PropTypes.func.isRequired
                    id="your_unique_id" // PropTypes.string.isRequired
                    numberOfMonths={1}
                    isOutsideRange={(day)=> false}
                    />
                    <textarea 
                    placeholder="notes for your expense (optional)"
                    rows="3"
                    cols="100"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    ></textarea>

                    <button>
                    Add
                    </button>
                </form>
            </div>

        )
    }
}