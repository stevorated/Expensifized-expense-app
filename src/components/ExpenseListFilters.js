import React from 'react'
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'

import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {
    state = {
        calanderFocused: null
    }

    onSelectFilter = (e)=>{
        if(e.target.value === 'date'){
            this.props.sortByDate()
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount()
        }
    }

    onTextChange = (e)=>{
        this.props.setTextFilter(e.target.value)
        console.log(e.target.value)
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onFocusChange = (calanderFocused) => {
        this.setState(()=> ({calanderFocused}))
    }

    render() {
        return (
            <div>
                <input 
                type="text"
                value={this.props.filters.text}
                onChange={this.onTextChange} 
                />
                <select 
                name="" 
                id="" 
                value={this.props.filters.sortBy}
                onChange={this.onSelectFilter}
                >
                
                <option value="date">Date</option>
                <option value="amount">Amount</option>
                </select>
                <DateRangePicker
                startDateId="datePickerStart"
                endDateId="datePickerEnd"
                startDate={this.props.filters.startDate}
                endDate={this.props.filters.endDate}
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calanderFocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={()=>false}
                showClearDates={true}
                />
            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch, props) => ({
    sortByDate: ()=> dispatch(sortByDate()),
    sortByAmount: ()=> dispatch(sortByAmount()),
    setTextFilter: (text)=> dispatch(setTextFilter(text)),
    setStartDate: (date)=> dispatch(setStartDate(date)),
    setEndDate: (date)=> dispatch(setEndDate(date)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)