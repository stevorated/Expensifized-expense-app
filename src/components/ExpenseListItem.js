import React from 'react'
import { connect } from 'react-redux'
import { Link } from "react-router-dom"
import moment from 'moment'



export const ExpenseListItem  = ({ id, desc, amount, createdAt }) => {
    if(!id) {
        return ( <p>No item</p> )
    }
    return (
        <div>
            <Link to={`/edit/${id}`}><h3>{desc}</h3></Link>
            <p>{amount/100} NIS - {moment(createdAt).format('LLL')}</p>
        </div>
    
    )
}


export default connect()(ExpenseListItem)