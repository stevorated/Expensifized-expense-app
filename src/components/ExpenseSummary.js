import React from 'react'
import { connect } from 'react-redux'
import numeral from 'numeral'
import selectExpensetotal from '../selectors/expenseTotal.selector'
import selectExpenses from '../selectors/expenses.selector'

export const ExpenseSummary = ({expenseCount, expenseTotal}) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const formattedExpenses = numeral(expenseTotal/100).format('$0, 0.00')
    return (
        <div>
            <h1>Viewing {expenseCount} {expenseWord} totalling {formattedExpenses}</h1>
        </div>
    )
}


const mapStateToProps = (state)=> {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: visibleExpenses.length,
        expenseTotal: selectExpensetotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)