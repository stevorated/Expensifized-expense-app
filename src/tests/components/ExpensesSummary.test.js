import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseSummary } from '../../components/ExpenseSummary'

test('sould correctly render ExpenseSummary with 1 item', ()=> {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expenseTotal={235} />)
    expect(wrapper).toMatchSnapshot()
})

test('sould correctly render ExpenseSummary with 1 item', ()=> {
    const wrapper = shallow(<ExpenseSummary expenseCount={23} expenseTotal={23523212} />)
    expect(wrapper).toMatchSnapshot()
})