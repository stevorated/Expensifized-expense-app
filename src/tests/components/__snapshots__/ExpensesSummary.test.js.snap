// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sould correctly render ExpenseSummary with 1 item 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;

exports[`sould correctly render ExpenseSummary with 1 item 2`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $235,232.12
  </h1>
</div>
`;
